atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/GetCandidateCellsForSweepingShim.java

```diff
@@ -0,0 +1,212 @@
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.sweep;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.commons.lang3.mutable.MutableLong;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.Lists;
+import com.google.common.collect.PeekingIterator;
+import com.google.common.collect.Sets;
+import com.google.common.io.Closer;
+import com.google.common.primitives.Longs;
+import com.google.common.primitives.UnsignedBytes;
+import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweeping;
+import com.palantir.atlasdb.keyvalue.api.CandidateCellForSweepingRequest;
+import com.palantir.atlasdb.keyvalue.api.Cell;
+import com.palantir.atlasdb.keyvalue.api.ImmutableCandidateCellForSweeping;
+import com.palantir.atlasdb.keyvalue.api.KeyValueService;
+import com.palantir.atlasdb.keyvalue.api.RangeRequest;
+import com.palantir.atlasdb.keyvalue.api.RowResult;
+import com.palantir.atlasdb.keyvalue.api.TableReference;
+import com.palantir.atlasdb.keyvalue.api.Value;
+import com.palantir.common.base.ClosableIterator;
+import com.palantir.common.base.ClosableIterators;
+
+// The plan is to make getCandidateCellsForSweeping() a method in KeyValueService.
+// This class provides a temporary implementation of the new call using existing KVS calls,
+// so that we can make this change in several smaller commits.
+public class GetCandidateCellsForSweepingShim {
+    private final KeyValueService keyValueService;
+
+    public GetCandidateCellsForSweepingShim(KeyValueService keyValueService) {
+        this.keyValueService = keyValueService;
+    }
+
+    /**
+     * For a given range of rows, returns all candidate cells for sweeping (and their timestamps).
+     * Here is the precise definition of a candidate cell:
+     * <blockquote>
+     *      Let {@code Ts} be {@code request.sweepTimestamp()}<br>
+     *      Let {@code Tu} be {@code request.minUncommittedTimestamp()}<br>
+     *      Let {@code V} be {@code request.shouldCheckIfLatestValueIsEmpty()}<br>
+     *      Let {@code Ti} be set of timestamps in {@code request.timestampsToIgnore()}<br>
+     *      <p>
+     *      Consider a cell {@code C}. Let {@code Tc} be the set of all timestamps for {@code C} that are strictly
+     *      less than {@code Ts}. Let {@code T} be {@code Tc \ Ti} (i.e. the cell timestamps minus the ignored
```

Comment:
Yeah that's the standard mathematical notation for the set difference: https://en.wikipedia.org/wiki/Complement_(set_theory)